xml_converter/tests/attribute_testing.py

```diff
@@ -0,0 +1,140 @@
+    "^The xml write function took [0-9]+ milliseconds to run$",
+    "^The protobuf write function took [0-9]+ milliseconds to run$"
+]
+
+
+if __name__ == "__main__":
+    # TODO: This will need to be extracted as more attributes and tests are added
+    attribute: Attribute_Tests = Attribute_Tests(
+        attribute_name="can_fade",
+        attribute_path="../test_cases/can_fade_tests",
+        tests=["_is_correct", "_is_different_file", "_is_wrong_output"],
+        output_result={
+            "_is_correct": ['Loading taco pack ../test_cases/can_fade_tests/can_fade_is_correct.xml'],
+            "_is_different_file": ['Loading taco pack ../test_cases/can_fade_tests/can_fade_is_different_file.xml'],
+            "_is_wrong_output": ['WRONG']
+        },
+        expect_output_to_be_equal={
+            "_is_correct": True,
+            "_is_different_file": True,
+            "_is_wrong_output": False,
+        },
+        expect_files_to_be_equal={
+            "_is_correct": True,
+            "_is_different_file": False,
+            "_is_wrong_output": True,
+        }
+    )
+
+    # Ensure that the test output directory is created
+    if not os.path.exists(attribute.attribute_path + "-output/"):
+        try:
+            os.makedirs(attribute.attribute_path + "-output/")
+        except OSError as e:
+            print(f"Error: {e}")
+
+    for test in attribute.tests:
+        arg1 = attribute.attribute_path + "/" + attribute.attribute_name + test + ".xml"
+        arg2 = attribute.attribute_path + "-output/" + attribute.attribute_name + test + ".xml"
+
+        result = run_xml_converter(arg_input_xml, arg1, arg_output_xml, arg2)
+        if result is not None:
+            # Remove lines about how long it took to run
+            split_result = result.split("\n")
+            filtered_array = []
+            for line in split_result:
+                match_found = False
+                for pattern in chrono_patterns:
+                    if re.search(pattern, line) or line == "":
+                        match_found = True
+                        break
+                if not match_found:
+                    filtered_array.append(line)
+
+            output_test: bool = True
+            file_test: bool = True
+
+            if are_arrays_equal(filtered_array, attribute.output_result[test]) != attribute.expect_output_to_be_equal[test]:
+                print(f"Output did not match for test {attribute.attribute_name}{test}")
+                print(f"Expected {attribute.output_result[test]} and got {filtered_array}")
+                output_test = False
+
+            if are_files_equal(arg1, arg2) != attribute.expect_files_to_be_equal[test]:
+                print(f"Files were incorrect for test {attribute.attribute_name}{test}")
+                print(f"Expected {attribute.expect_files_to_be_equal[test]}")
+                print(f"arg1: {arg1}")
+                print(f"arg2: {arg2}")
+                file_test = False
+
```

Comment:
Not necessarily required for either of these things but in general if there is an error because things are different then we will want to know what part of them is different. We can ignore this for now because it is not critical but it will be a useful thing to have later.

Reply:
I found a module that can generate diffs between files so I have switched to using that.